### PR TITLE
Adding SMURFF recipe + dependencies

### PR DESCRIPTION
**Pull request by tvandera**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by tvandera** [comment]
FYI, I already used smithy to make feedstocks - just for testing.
- h5sparse-tensor: https://github.com/tvandera/h5sparse-tensor-feedstock
- smurff: https://github.com/tvandera/smurff-feedstock

The feedstocks build successfully on Azure: https://dev.azure.com/tomvanderaa/feedstock-builds

**Comment by tvandera** [comment]
Yeah! 🎉

**Comment by tvandera** [comment]
@conda-forge/staged-recipes Ready to be merged.

File: recipes/h5sparse-tensor/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    - python -m unittest
+
+about:
+  home: "https://github.com/tvandera/h5sparse"
+  license: MIT
+  license_family: MIT
+  license_file: LICENSE
+  summary: "Scipy sparse matrix in HDF5."
+  doc_url: 
+  dev_url:
```
Comment (xhochy) [inline recipes/h5sparse-tensor/meta.yaml:48]: Please omit empty sections
```suggestion
```

File: recipes/h5sparse-tensor/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    - scipy
+    - six
+    - setuptools
+
+test:
+  imports:
+    - h5sparse
+  files:
+    - tests.py
+  commands:
```
Comment (xhochy) [inline recipes/h5sparse-tensor/meta.yaml:38]:
```suggestion
  requires:
    - pip
  commands:
    - pip check
```

File: recipes/h5sparse-tensor/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+requirements:
+  host:
+    - h5py
+    - numpy
+    - python >=3.6
+    - scipy
+    - six
+    - pip
+    - nose
```
Comment (xhochy) [inline recipes/h5sparse-tensor/meta.yaml:24]: This should be sufficient for host
```suggestion
  host:
    - python >=3.6
    - pip
```

File: recipes/smurff/build.sh [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+export CMAKE_ARGS="-DENABLE_MKL=ON -DCMAKE_INSTALL_PREFIX=$PREFIX -DENABLE_MPI=OFF"
```
Comment (xhochy) [inline recipes/smurff/build.sh:3]: Extend the existing `CMAKE_ARGS`
```suggestion
export CMAKE_ARGS="${CMAKE_ARGS} -DENABLE_MKL=ON -DCMAKE_INSTALL_PREFIX=$PREFIX -DENABLE_MPI=OFF"
```

File: recipes/smurff/build.sh [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+# make sure we use CONDA_BUILD_SYSROOT
+# https://github.com/conda/conda-build/issues/3452#issuecomment-47539707
+[[ ${target_platform} == "osx-64" ]] && \
+    CMAKE_ARGS="$CMAKE_ARGS -DCMAKE_OSX_SYSROOT=${CONDA_BUILD_SYSROOT}"
```
Comment (xhochy) [inline recipes/smurff/build.sh:8]: Shouldn't be needed with the above `CMAKE_ARGS` extension.
```suggestion
```

File: recipes/smurff/conda_build_config.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+hdf5:
```
Comment (xhochy) [inline recipes/smurff/conda_build_config.yaml:1]: Please remove this file, this (the pinnings) shouldn't be needed when build inside of the conda-forge infrastructure.

File: recipes/smurff/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  build:
+    - cmake
+    - {{ compiler('cxx') }}
+    - {{ compiler('c') }}
+    - llvm-openmp  # [osx]
+    - libgomp      # [linux]
+  host:
+    - mkl-devel
+    - eigen
+    - libboost
```
Comment (xhochy) [inline recipes/smurff/meta.yaml:26]:
```suggestion
    - boost-cpp
```

File: recipes/smurff/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    - python {{ python }}
+    - setuptools
+    - setuptools_scm
+    - numpy
+  run:
+    - python {{ python }}
+    - mkl
+    - llvm-openmp  # [osx]
+    - libgomp      # [linux]
+    - libboost     # [not win]
```
Comment (xhochy) [inline recipes/smurff/meta.yaml:39]: These should be added by their `run_exports` and `boost-cpp` is the correct name here.
```suggestion
  - boost-cpp
```

File: recipes/smurff/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+requirements:
+  build:
+    - cmake
+    - {{ compiler('cxx') }}
+    - {{ compiler('c') }}
+    - llvm-openmp  # [osx]
+    - libgomp      # [linux]
+  host:
+    - mkl-devel
```
Comment (isuruf) [inline recipes/smurff/meta.yaml:24]: It looks like `mkl` is not strictly necessary, but a blas/lapack library is. Please see conda-forge docs on how to use blas libraries.

File: recipes/smurff/conda_build_config.yaml (hunk 6 above)
Comment (tvandera) [reply recipes/smurff/conda_build_config.yaml]: File removed.

**Comment by tvandera** [comment]
Hi folks, thanks for all your suggestions. I've been able to implement most of them, except for these two which need upstream changes:

- *h5sparse-tensor*: nose is still a host requirement even though we use unittests. Need to change upstream for this AFAIK.
- *smurff*: using plain blas/lapack on windows does not seem to work due to lack of support of `_Complex` in MSVC (see [this post](https://stackoverflow.com/questions/24853450/errors-using-lapack-c-header-in-c-with-visual-studio-2010)). I've spilt up requirements: MKL on Windows, blas/lapack on OSX/Linux.

Is it ok to leave those for a next release in upstream (which I also manage)? If so, @conda-forge-admin, please ping team again.

T.

**Comment by isuruf** [comment]
> smurff: using plain blas/lapack on windows does not seem to work due to lack of support of _Complex in MSVC (see this post). I've spilt up requirements: MKL on Windows, blas/lapack on OSX/Linux.
 
 Can you try adding `-DLAPACK_COMPLEX_STRUCTURE=1` to c compilation flags?

**Comment by tvandera** [comment]
@isuruf How are blas implementations selected during feedstock builds?

For some reason my own feedstock build selects mkl on Windows  (See Azure Pipelines logs [here]( https://dev.azure.com/tomvanderaa/feedstock-builds/_build/results?buildId=46&view=logs&j=171a126d-c574-5c8c-1269-ff3b989e923d&t=1183ba29-a0b5-5324-8463-2a49ace9e213&l=53)) while in staged-recipes netlib is selected (See Azure Pipelines logs [here](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=360061&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=23010c01-6a9d-5f94-9037-792618644b77&l=993))

**Comment by tvandera** [comment]
> > smurff: using plain blas/lapack on windows does not seem to work due to lack of support of _Complex in MSVC (see this post). I've spilt up requirements: MKL on Windows, blas/lapack on OSX/Linux.
> 
> Can you try adding `-DLAPACK_COMPLEX_STRUCTURE=1` to c compilation flags?

@isuruf The following seems to work ([MSVC docs](https://docs.microsoft.com/en-us/cpp/c-runtime-library/complex-math-support))

```
+  if (MSVC)
+    # needed because MSVC does not have support for c-type _Complex
+    add_definitions(-Dlapack_complex_float=_Fcomplex -Dlapack_complex_double=_Dcomplex)
+  endif()
```

It's currently included as a patch in recipes/smurff/ and upstream for next release.

**Comment by tvandera** [comment]
@conda-forge-admin, please ping team

**Comment by tvandera** [comment]
@conda-forge-admin, please ping team. Has been ready for a while! Please merge :)

**Comment by tvandera** [comment]
Ready for merge, but no reply from conda-forge admins - giving up - closing PR.

**Pull request closed without merging**